src/python/pants/backend/jvm/subsystems/jvm_tool_mixin.py

```diff
@@ -145,6 +150,72 @@ def formulate_help():
     jvm_tool = cls.JvmTool(register.scope, key, classpath, main, custom_rules)
     JvmToolMixin._jvm_tools.append(jvm_tool)
 
+  # TODO: Deprecate .register_jvm_tool()?
+  class JvmToolDeclaration(datatype([
+      ('tool_name', text_type),
+      # If the tool's `main` class name is supplied the tool classpath will be shaded!
+      ('main', Exactly(text_type, type(None))),
+      ('classpath', tuple),
+      ('custom_rules', tuple),
+  ])):
+    """A mostly-typed specification for a JVM tool. Can be passed around across python modules."""
+
+    def __new__(cls, tool_name, main=None, classpath=None, custom_rules=None):
+      return super(JvmToolMixin.JvmToolDeclaration, cls).__new__(
+        cls, text_type(tool_name), main, tuple(classpath or []), tuple(custom_rules or []))
+
+  class JvmToolDeclError(Exception): pass
+
+  @classmethod
+  def register_jvm_tool_decl(cls, register, decl, **kwargs):
+    """Register a `JvmToolDeclaration` with `.register_jvm_tool()`."""
+    if not isinstance(decl, cls.JvmToolDeclaration):
+      raise cls.JvmToolDeclError(
+        'decl {} must be an instance of {}.{}'
+        .format(decl, cls.__name__, cls.JvmToolDeclaration.__name__))
+    kwargs = dict(
+      classpath=list(decl.classpath),
+      custom_rules=list(decl.custom_rules)
+    )
+    if decl.main:
+      kwargs['main'] = decl.main
+    cls.register_jvm_tool(register,
+                          decl.tool_name,
+                          **kwargs)
+
+  class CombinedJvmToolsError(JvmToolDeclError): pass
+
+  @classproperty
+  def combined_jvm_tool_names(cls):
```

Comment:
It looks like this is only used in `ensure_combined_jvm_tool_classpath`... could it just be an argument instead?

Reply:
Yes, that would make this a lot easier to read.

Reply:
Done!